exercise/static/exercise/chapter.js

```diff
@@ -561,19 +543,28 @@
 					});
 				});
 			} else {
-				chapter.openModal(chapter.messages.submit);
 				var exercise = this;
+				var submission_complete = false;
+				if (exercise.quiz) {
+					setTimeout(function(){
```

Comment:
What happens if the submission is not completed after 1.5s? Timers are typically the wrong solution for everything (sometimes they are the right, but typically not). You can do same in the jqeury ajax `.always` callback.

Additionally, global variables, or variables which are conceptually polled (like here), do not fit in callback/event/future based programs (see https://en.wikipedia.org/wiki/Event-driven_programming for concepts).

As a rule of thumb, if you use `var` in JavaScript, then there is basically always a better way to do it. Bascally, all "variables" should be stored in `const`.

Reply:
As I (tried at least) to explain above is the process what happens here:

1. User clicks submit, and the "clock" on 1,5 s is started, submitAjax is called
2. For the first 1,5 s, everything in the form is shown as it is, except the submit-button is disabled
3. After 1,5 s has passed, and the process is stuck in these retrys https://github.com/apluslms/a-plus/blob/master/exercise/static/exercise/chapter.js#L430 the progress-bar of the images above is shown, it says "posting submission".
4a. If the submitAjax fails, the "clock" is stopped, the progress bar hidden an this current message is shown:
![image](https://user-images.githubusercontent.com/43036333/90012745-a80aff00-dcac-11ea-9a5d-efb8196a472e.png).
4b. Alternatively, if the submitAjax succeeds on some of the retrys, the progress bar is hidden and the result form is shown.

So the progress bar was there just to indicate to the user that "hold on, this submission is taking a bit longer than expected". And since the quiz submissions tend to be fast, I thought it does not make sense to show the user just a quick flash of a progress bar, hence the 1,5 seconds before the progress bar is changed. That was at least I was trying to accomplish here. 

What do you think? What would you differently here?

Reply:
What happens when submission takes 10 seconds? Let's assume the user press "send" and after 1.5s they change some values and press send again (or don't). After 10 seconds, the JS would replace the form with a new feedback and all the students edits would be deleted?

I think this was the reason why I originally were thinking of an overlay to block the user from editing the fields while the data is under javascript control and might be replaced.

Reply:
These are valid points! 

However, originally [in the issue](https://github.com/apluslms/a-plus/issues/493), there was only talking about removing the empty modal. So the overlay-idea is new to me, at least I am not able to recall a situation where it would have come up. I can only recall talking about presenting the content of the modal within the exercise frame. It also differs from this idea:

> For now, I was mostly interested about the communication error, which changes the text in the progress bar to "communication error", but so far that hasn't been very visible. I think that could be done better here. For example, the progress bar could be removed, the submission button enabled, and the error message rendered there (which is removed if the person presses submit again).

But. I still have some time to implement this and I would like to get it right :) 

So could you clarify what would you like to be implemented here regarding the overlay? Is what you mean by the "data being under javascript control" that the overlay would be launched when the student hits "submit" and disappears when one of the events "exercise-submission-failure", "submission-finished" or "submission-aborted" is introduced? So similar kind of flow as with the submit-button now: 
https://github.com/apluslms/a-plus/blob/master/exercise/static/exercise/chapter.js#L857

Just so I know on how to continue.

Reply:
Basically that. We can chat about it on monday!

Reply:
Since the `setTimeout` and `var submission_complete` are still here, does it mean that you ran out of time for doing this without them? Just to clarify, no panic if that's the case :)

Reply:
Yes, you are correct :) I didn't come up with a good way to replace them without using `setTimeout,` since the status of the submission does not actually change (I think). And it also looked a bit weird to have the overlay flashing there for a while with fast submissions. If you have any other ideas how that could be accomplished, or if you'd like 1s-delay to be removed all in all and the overlay to flash there, even for a split-second, just do let me know! That can still easily be changed of course :)

Reply:
You could use a transparent overlay. I think the CSS can be used to change the transparency e.g. in 0.4s, thus if it flashes, then it's not actually visible to the user, but does still block any input.

Reply:
Uuu, that's a very nice idea! Thanks a lot!! I will try that today :)

Reply:
I tried the transition from transparent background to the darked background, it was a very good idea. But then the progress bar looked a bit weird, "hanging in the air" if someone happens to look at the top part. 

I also tried to just change the opacity of the background, but the child-elements also have then the same opacity and so the progress bar ended up looking weird. And there was no good way around that child-element-problem.

But talking about the transparency gave me another idea, and I implemented that. So I added the progress-bar within the exercise-frame, but in the upper part and the overlay as a a shadowy gray. 

It might not before the perfect solution, since there is that slight twitch of showing the progress bar, when the submission is fast. However, I guess it is also somehow expected behavior from a form, that the elements might move around slightly upon submission. And at least the input is now blocked by the overlay, so students don't go writing a lot of stuff there and lose it if the connection fails. 

And in any case, I think now this already significantly better than the modal. 

The JS is not the prettiest still cause there are those branches for different exercises. But this could be set as a base for removing the modal in the future from other exercises, and then those branches wouldn't be no longer needed.